### PR TITLE
CSMA-CA procedure

### PR DESCRIPTION
**Pull request by hubertmis**
This PR provides initial support for CSMA-CA procedure. Limitation of this implementation is that regular transmission `nrf_drv_radio802154_transmit()` or `nrf_drv_radio802154_transmit_raw()` should not be used when CSMA-CA is enabled. There is noted TODO to fix this limitation.

### CODE REVIEW

**Comment by KaSroka** [comment]
Regression test

**Comment by hubertmis** [comment]
Regression test

**Comment by hubertmis** [comment]
Regression test

**Comment by KaSroka** [comment]
Regression test

**Comment by KaSroka** [comment]
Regression test

File: src/nrf_drv_radio802154.h [hunk 1]
```diff
@@ -761,6 +716,46 @@ void nrf_drv_radio802154_cca_cfg_set(const nrf_drv_radio802154_cca_cfg_t * p_cca
...
+#if NRF_DRV_RADIO802154_CSMA_CA_ENABLED
+
+/**
+ * @brief Perform CSMA-CA procedure and transmit frame in case of success.
+ *
+ * End of CSMA-CA procedure is notified by @sa nrf_drv_radio802154_transmitted() or
+ * @sa nrf_drv_radio802154_transmit_failed() function.
+ *
+ * @note CSMA-CA procedure does not time out waiting for ACK frame in case frame with ACK request
+ *       bit set was transmitted. To stop waiting for ACK frame, MAC layer should use
```
Comment (LuDuda) [inline src/nrf_drv_radio802154.h]: To stop -> To start?

File: src/nrf_drv_radio802154.h [hunk 2]
```diff
@@ -761,6 +716,46 @@ void nrf_drv_radio802154_cca_cfg_set(const nrf_drv_radio802154_cca_cfg_t * p_cca
...
+void nrf_drv_radio802154_transmit_csma_ca_raw(const uint8_t *p_data);
+
+/**
+ * @brief Perform CSMA-CA procedure and transmit frame in case of success.
+ *
+ * End of CSMA-CA procedure is notified by @sa nrf_drv_radio802154_transmitted() or
+ * @sa nrf_drv_radio802154_transmit_failed() function.
+ *
+ * @note CSMA-CA procedure does not time out waiting for ACK frame in case frame with ACK request
+ *       bit set was transmitted. To stop waiting for ACK frame, MAC layer should use
```
Comment (LuDuda) [inline src/nrf_drv_radio802154.h]: To stop -> To start?

File: src/nrf_drv_radio802154_fsm_hooks.c [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#include "nrf_drv_radio802154_types.h"
+
+#include "mac_features/nrf_drv_radio802154_csmaca.h"
+
+typedef bool (* tx_failed_hook)(nrf_drv_radio802154_tx_error_t error);
+
+static const tx_failed_hook m_tx_failed_hooks[] =
+{
+#if NRF_DRV_RADIO802154_CSMA_CA_ENABLED
+        nrf_drv_radio802154_csmaca_tx_failed_hook,
```
Comment (LuDuda) [inline src/nrf_drv_radio802154_fsm_hooks.c]: nit: seems like too many spaces before.

File: src/mac_features/nrf_drv_radio802154_csmaca.c [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        nrf_drv_radio802154_notify_transmit_failed(NRF_DRV_RADIO802154_TX_ERROR_BUSY_CHANNEL);
+    }
+}
+
+/**
+ * @brief Delay CCA procedure for random (2^BE - 1) unit backoff periods.
+ */
+static void random_backoff_start(void)
+{
+    uint8_t backoff_periods = rand() % (1 << m_be);
```
[the remark this answers is not in the document]
Comment (hubertmis) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: I don't think we need hardware generated random number for backoff periods randomization. It's not security issue. `rand()` returns pseudorandom number.

File: src/mac_features/nrf_drv_radio802154_csmaca.c [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    mp_psdu = p_data;
+    m_nb    = 0;
+    m_be    = NRF_DRV_RADIO802154_CSMA_CA_MIN_BE;
+
+    random_backoff_start();
+}
+
+bool nrf_drv_radio802154_csmaca_tx_failed_hook(nrf_drv_radio802154_tx_error_t error)
+{
+    (void)error;
```
[the remark this answers is not in the document]
Comment (hubertmis) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: FCS casting example shows it without space. Is there any reason to add space here?

File: src/nrf_drv_radio802154.h (hunk 1 above)
Comment (hubertmis) [reply src/nrf_drv_radio802154.h]: I'll rephrase this sentence

File: src/mac_features/nrf_drv_radio802154_csmaca.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (KaSroka) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: And as far as I remember we have true random seed (or not?)

File: src/mac_features/nrf_drv_radio802154_csmaca.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (LuDuda) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: My preferation, and for consistency, if you don't like it, you should remove extra space from line 85.

**Comment by KaSroka** [comment]
Regression test

File: src/mac_features/nrf_drv_radio802154_csmaca.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (LuDuda) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: Do we? Don't we have the same value of rand() after reboots?

[the remark this answers is not in the document]
Comment (hubertmis) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: I would leave setting random seed to application. We could add info in documentation that application should use `srand()` before CSMA-CA. Would it be acceptable?

File: src/mac_features/nrf_drv_radio802154_csmaca.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (hubertmis) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: I'll update 85 to stick with FCS

File: src/mac_features/nrf_drv_radio802154_csmaca.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (LuDuda) [reply src/mac_features/nrf_drv_radio802154_csmaca.c]: Yes.

File: src/nrf_drv_radio802154.c [hunk 6]
```diff
@@ -400,6 +419,26 @@ void nrf_drv_radio802154_cca_cfg_get(nrf_drv_radio802154_cca_cfg_t * p_cca_cfg)
     nrf_drv_radio802154_pib_cca_cfg_get(p_cca_cfg);
 }
 
+#if NRF_DRV_RADIO802154_CSMA_CA_ENABLED
+
+void nrf_drv_radio802154_transmit_csma_ca_raw(const uint8_t *p_data)
```
[the remark this answers is not in the document]
Comment (hubertmis) [reply src/nrf_drv_radio802154.c]: I'll use csma_ca - seems easier to read.

File: src/nrf_drv_radio802154_config.h [hunk 7]
```diff
@@ -254,6 +254,56 @@ extern "C" {
...
+ *
+ */
+#ifndef NRF_DRV_RADIO802154_CSMA_CA_MIN_BE
+#define NRF_DRV_RADIO802154_CSMA_CA_MIN_BE 3
+#endif
+
+/**
+ * @def NRF_DRV_RADIO802154_CSMA_CA_MAX_BE
+ *
+ * The maximum value of the backoff exponent, BE, in the CSMA-CA algorithm.
```
[the remark this answers is not in the document]
Comment (hubertmis) [reply src/nrf_drv_radio802154_config.h]: It's copy from the spec. Should I unify it? Do you prefer one version?

[the remark this answers is not in the document]
Comment (rlubos) [reply src/nrf_drv_radio802154_config.h]: Brackets look better IMHO, but thats not the case - my intention was to keep single approach. But if it's a copy-paste from the specification, and you want to keep it unmodified, I'm fine with it.

[the remark this answers is not in the document]
Comment (hubertmis) [reply src/nrf_drv_radio802154_config.h]: I'll leave it as it is then.

**Comment by hubertmis** [comment]
Fixed all comments. Ready for 2nd review iteration.

**Pull request merged**